File: src/features/providers/newChat.ts
import { z } from "zod";
import { PROVIDERS, PROVIDER_ORDER, type Provider } from "./model";

/**
 * How a placement should reach a blank conversation.
 *
 * The two mechanisms fail in different ways, which is why the user picks
 * between them instead of getting one hard-coded choice. Clicking the page's
 * own control is instant and keeps the app mounted, but depends on that
 * page's markup. Navigating to a new-chat address always works, but pays a
 * full reload. `auto` tries the click first and navigates only if it fails.
 */
export const NEW_CHAT_MODES = ["auto", "button", "url", "off"] as const;
export type NewChatMode = (typeof NEW_CHAT_MODES)[number];
export const DEFAULT_NEW_CHAT_MODE: NewChatMode = "auto";

export const NEW_CHAT_MODE_LABELS: Record<
  NewChatMode,
  { title: string; detail: string }
> = {
  auto: {
    title: "Automatic (recommended)",
    detail: "Press the page's New chat button, and reload the page if that fails.",
  },
  button: {
    title: "Button only",
    detail: "Press the page's New chat button. Never reloads the page.",
  },
  url: {
    title: "Address only",
    detail: "Always open the new chat address. Slower, but does not depend on the page.",
  },
  off: {
    title: "Off",
    detail: "Keep the conversation that is already open.",
  },
};

/**
 * Mirrors the limits enforced in `src-tauri/src/provider/new_chat.rs`. The
 * native side is authoritative; these exist so a bad paste is explained in the
 * form rather than rejected after the user saves it.
 */
const MAX_SIGNAL_BYTES = 256;
const MAX_ATTRIBUTES = 8;
const MAX_URL_BYTES = 2_048;
const MAX_PASTE_BYTES = 64 * 1024;

const ALLOWED_PLAIN_ATTRIBUTES = ["href", "id", "name", "role", "type"];

function byteLength(value: string): number {
  return new TextEncoder().encode(value).length;
}

const BoundedSignalSchema = z
  .string()
  .trim()
  .min(1)
  .refine((value) => byteLength(value) <= MAX_SIGNAL_BYTES)
  .refine((value) => !/\p{Cc}/u.test(value));

export const NewChatMatcherSchema = z
  .strictObject({
    testId: BoundedSignalSchema.optional(),
    label: BoundedSignalSchema.optional(),
    href: BoundedSignalSchema.optional(),
    attributes: z
      .array(
        z.strictObject({
          name: z
            .string()
            .trim()
            .min(1)
            .refine((name) => isAllowedAttributeName(name.toLowerCase())),
          value: z
            .string()
            .trim()
            .refine((value) => byteLength(value) <= MAX_SIGNAL_BYTES)
            .refine((value) => !/\p{Cc}/u.test(value)),
        }),
      )
      .max(MAX_ATTRIBUTES)
      .optional(),
  })
  .refine(
    (matcher) =>
      matcher.testId !== undefined ||
      matcher.label !== undefined ||
      matcher.href !== undefined ||
      (matcher.attributes?.length ?? 0) > 0,
    { message: "The description has nothing to match on." },
  );

export type NewChatMatcher = z.infer<typeof NewChatMatcherSchema>;

export const NewChatProviderOverrideSchema = z.strictObject({
  url: z.string().trim().min(1).optional(),
  matcher: NewChatMatcherSchema.optional(),
});

export type NewChatProviderOverride = z.infer<
  typeof NewChatProviderOverrideSchema
>;

export type NewChatOverrides = Partial<
  Record<Provider, NewChatProviderOverride>
>;

/**
 * Same shape rule as the native allowlist: the `data-`/`aria-` families plus a
 * short list of identity attributes. `class` and `style` are absent by
 * design — they describe appearance and change on nearly every deploy.
 */
export function isAllowedAttributeName(name: string): boolean {
  if (!/^[a-z0-9_-]+$/.test(name)) return false;
  if (name.startsWith("data-") || name.startsWith("aria-")) {
    return name.length > "data-".length;
  }
  return ALLOWED_PLAIN_ATTRIBUTES.includes(name);
}

// ---------------------------------------------------------------------------
// Reading settings back
// ---------------------------------------------------------------------------

export function decodeStoredNewChatMode(value: unknown): NewChatMode {
  return NEW_CHAT_MODES.includes(value as NewChatMode)
    ? (value as NewChatMode)
    : DEFAULT_NEW_CHAT_MODE;
}

/**
 * A damaged or half-written override must not cost the user the whole
 * setting, so each provider is validated on its own and a bad one is dropped.
 */
export function decodeStoredNewChatOverrides(value: unknown): NewChatOverrides {
  if (typeof value !== "object" || value === null) return {};

  const overrides: NewChatOverrides = {};
  for (const provider of PROVIDER_ORDER) {
    const candidate = (value as Record<string, unknown>)[provider];
    if (candidate === undefined) continue;

    const result = NewChatProviderOverrideSchema.safeParse(candidate);
    if (!result.success) continue;
    if (result.data.url === undefined && result.data.matcher === undefined) {
      continue;
    }
    // An address that no longer passes validation would be rejected natively
    // on every placement, so it is dropped here and the built-in one is used.
    if (
      result.data.url !== undefined &&
      validateNewChatUrl(provider, result.data.url) !== null
    ) {
      if (result.data.matcher === undefined) continue;
      overrides[provider] = { matcher: result.data.matcher };
      continue;
    }
    overrides[provider] = result.data;
  }
  return overrides;
}

export function newChatUrlFor(
  provider: Provider,
  overrides: NewChatOverrides,
): string | undefined {
  return overrides[provider]?.url;
}

export function newChatMatcherFor(
  provider: Provider,
  overrides: NewChatOverrides,
): NewChatMatcher | undefined {
  return overrides[provider]?.matcher;
}

/**
 * Plain-language list of what a saved matcher looks for, so the setting can be
 * checked at a glance long after it was pasted.
 */
export function describeNewChatMatcher(matcher: NewChatMatcher): string[] {
  const signals: string[] = [];
  if (matcher.testId) signals.push(`test id "${matcher.testId}"`);
  if (matcher.label) signals.push(`name "${matcher.label}"`);
  if (matcher.href) signals.push(`link "${matcher.href}"`);
  for (const { name, value } of matcher.attributes ?? []) {
    signals.push(value === "" ? name : `${name}="${value}"`);
  }
  return signals;
}

// ---------------------------------------------------------------------------
// Address validation
// ---------------------------------------------------------------------------

/**
 * Returns a message explaining why the address is unusable, or `null` when it
 * is fine. The rules match `resolve_new_chat_url` on the native side: a
 * placement refuses to fill a page that is not on the provider's own host, so
 * an address that leaves the origin would strand the pane somewhere the prompt
 * can never land.
 */
export function validateNewChatUrl(
  provider: Provider,
  candidate: string,
): string | null {
  const trimmed = candidate.trim();
  if (!trimmed) return "Enter an address, or leave the field empty.";
  if (byteLength(trimmed) > MAX_URL_BYTES) return "That address is too long.";

  let url: URL;
  try {
    url = new URL(trimmed);
  } catch {
    return "That is not a valid web address.";
  }

  if (url.protocol !== "https:") return "The address must start with https://.";
  if (url.username || url.password) {
    return "The address must not contain a username or password.";
  }
  if (url.port !== "" && url.port !== "443") {
    return "The address must use the standard https port.";
  }

  const { host, label } = PROVIDERS[provider];
  if (url.hostname !== host) {
    return `The address must stay on ${host}, so it must look like https://${host}/… for ${label}.`;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Reading a pasted element
// ---------------------------------------------------------------------------

/**
 * Attributes that describe what a control is doing right now rather than
 * which control it is. Matching on them makes an override rot faster than
 * having no override at all, so they are dropped from a paste.
 */
const VOLATILE_ATTRIBUTE_NAMES = new Set([
  "aria-activedescendant",
  "aria-busy",
  "aria-checked",
  "aria-controls",
  "aria-current",
  "aria-describedby",
  "aria-disabled",
  "aria-expanded",
  "aria-haspopup",
  "aria-hidden",
  "aria-labelledby",
  "aria-live",
  "aria-owns",
  "aria-pressed",
  "aria-selected",
  "aria-valuenow",
  "data-active",
  "data-checked",
  "data-disabled",
  "data-focus",
  "data-focused",
  "data-highlighted",
  "data-hover",
  "data-index",
  "data-key",
  "data-loading",
  "data-open",
  "data-pressed",
  "data-reactid",
  "data-reactroot",
  "data-revealed",
  "data-selected",
  "data-state",
  "data-value",
]);

/**
 * Framework bookkeeping. The names are generated per build, so a match today
 * says nothing about tomorrow.
 */
const VOLATILE_ATTRIBUTE_PREFIXES = [
  "data-headlessui-",
  "data-ng-",
  "data-radix-",
  "data-svelte-",
  "data-v-",
];

/**
 * `role` and `type` are structural HTML rather than identity: nearly every
 * clickable row on a provider page carries them, so matching on one would let
 * a sidebar conversation borrow the confidence meant for the real control.
 */
const GENERIC_ATTRIBUTE_NAMES = new Set(["role", "type"]);

/**
 * Values a build tool produced. An id like `:r7:` or `mat-button-14` is
 * reassigned on the next render, so it is worse than no signal at all.
 */
const GENERATED_VALUE_PATTERNS = [
  /^:[a-z0-9]+:$/i,
  /^[0-9a-f]{12,}$/i,
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-/i,
  /^(cdk|ember|headlessui|mat|mui|ng|radix)[-:]/i,
  /\d{4,}/,
];

function looksGenerated(value: string): boolean {
  return GENERATED_VALUE_PATTERNS.some((pattern) => pattern.test(value));
}

export type NewChatParseResult =
  | { ok: true; matcher: NewChatMatcher; signals: string[]; dropped: number }
  | { ok: false; message: string };

/**
 * Text as a person reads it. Keyboard hints and screen-reader-only decoration
 * are skipped so ChatGPT's "New chat ⇧⌘O" row reduces to "New chat" — the
 * same reduction the in-page scorer performs, so the two agree on a match.
 */
function visibleTextOf(element: Element): string {
  let text = "";
  const walk = (node: Node) => {
    if (text.length > MAX_SIGNAL_BYTES) return;
    if (node.nodeType === 3) {
      text += node.nodeValue ?? "";
      return;
    }
    if (node.nodeType !== 1) return;
    const child = node as Element;
    if (child.tagName === "KBD" || child.tagName === "SCRIPT") return;
    if (child.tagName === "STYLE" || child.tagName === "SVG") return;
    if (child.getAttribute("aria-hidden") === "true") return;
    child.childNodes.forEach(walk);
  };
  walk(element);
  return text.replace(/\s+/g, " ").trim();
}

/**
 * Kept as a path so a matcher recorded from an absolute link still matches the
 * relative href the page renders, and vice versa.
 */
function hrefSignalOf(element: Element): string | undefined {
  const raw = element.getAttribute("href")?.trim();
  if (!raw) return undefined;

  let path = raw;
  try {
    path = new URL(raw, "https://provider.invalid/").pathname;
  } catch {
    // A non-URL href is compared verbatim, exactly as the scorer would.
  }
  // A link to one specific conversation is the opposite of a new-chat control;
  // recording it would teach the matcher to reopen an old thread.
  if (path.split("/").some(looksGenerated)) return undefined;
  return path;
}

function attributeRank(name: string): number {
  if (name.startsWith("data-")) return 0;
  if (name === "name") return 2;
  if (name === "id") return 4;
  return 6;
}

/**
 * Turns a DevTools "Copy element" paste into a set of signals.
 *
 * The HTML is parsed into an inert document and never enters the live page:
 * nothing is fetched, no script runs, and only whitelisted attribute values
 * leave this function. What is returned is data that eventually reaches the
 * provider page as a JSON argument, never as script text.
 */
export function parseNewChatElement(html: string): NewChatParseResult {
  const source = html.trim();
  if (!source) {
    return { ok: false, message: "Paste the copied element first." };
  }
  if (byteLength(source) > MAX_PASTE_BYTES) {
    return {
      ok: false,
      message: "That paste is too large. Copy just the button element.",
    };
  }
  if (typeof DOMParser === "undefined") {
    return {
      ok: false,
      message: "Prompter could not read the pasted element on this system.",
    };
  }

  let element: Element | null;
  try {
    const document = new DOMParser().parseFromString(source, "text/html");
    element = document.body.firstElementChild;
  } catch {
    return { ok: false, message: "Prompter could not read that paste." };
  }
  if (!element) {
    return {
      ok: false,
      message:
        "That does not look like an HTML element. In the browser, right-click the New chat button, choose Inspect, then right-click the highlighted line and choose Copy → Copy element.",
    };
  }

  const signals: string[] = [];
  const matcher: {
    testId?: string;
    label?: string;
    href?: string;
    attributes: { name: string; value: string }[];
  } = { attributes: [] };

  const testId = (
    element.getAttribute("data-testid") ??
    element.getAttribute("data-test-id") ??
    ""
  ).trim();
  if (testId && byteLength(testId) <= MAX_SIGNAL_BYTES) {
    matcher.testId = testId;
    signals.push(`test id "${testId}"`);
  }

  const label = (element.getAttribute("aria-label") ?? "").trim() ||
    visibleTextOf(element);
  if (label && byteLength(label) <= MAX_SIGNAL_BYTES) {
    matcher.label = label;
    signals.push(`name "${label}"`);
  }

  const href = hrefSignalOf(element);
  if (href && byteLength(href) <= MAX_SIGNAL_BYTES) {
    matcher.href = href;
    signals.push(`link "${href}"`);
  }

  let dropped = 0;
  const usable: { name: string; value: string; rank: number }[] = [];
  for (const attribute of Array.from(element.attributes)) {
    const name = attribute.name.toLowerCase();
    if (name === "data-testid" || name === "data-test-id") continue;
    if (name === "aria-label" || name === "href") continue;

    const value = attribute.value.trim();
    if (
      !isAllowedAttributeName(name) ||
      GENERIC_ATTRIBUTE_NAMES.has(name) ||
      VOLATILE_ATTRIBUTE_NAMES.has(name) ||
      VOLATILE_ATTRIBUTE_PREFIXES.some((prefix) => name.startsWith(prefix)) ||
      byteLength(value) > MAX_SIGNAL_BYTES ||
      /\p{Cc}/u.test(value) ||
      (value !== "" && looksGenerated(value))
    ) {
      dropped += 1;
      continue;
    }
    // An empty value is still a signal: the scorer treats it as "this
    // attribute is present", which is how flag attributes are written.
    const isFlag = value === "" || value === "true" || value === "false";
    usable.push({ name, value, rank: attributeRank(name) + (isFlag ? 1 : 0) });
  }

  usable.sort((left, right) => left.rank - right.rank);
  for (const { name, value } of usable.slice(0, MAX_ATTRIBUTES)) {
    matcher.attributes.push({ name, value });
    signals.push(value === "" ? name : `${name}="${value}"`);
  }
  dropped += Math.max(0, usable.length - MAX_ATTRIBUTES);

  const parsed = NewChatMatcherSchema.safeParse(
    matcher.attributes.length > 0
      ? matcher
      : { testId: matcher.testId, label: matcher.label, href: matcher.href },
  );
  if (!parsed.success) {
    return {
      ok: false,
      message:
        "Prompter found nothing reliable to match in that element. Copy the New chat button itself rather than a wrapper around it.",
    };
  }

  return { ok: true, matcher: parsed.data, signals, dropped };
}

File: src-tauri/src/provider/new_chat.rs
//! Resetting a provider pane to a blank conversation.
//!
//! Two mechanisms exist, and they fail in different ways, which is why both
//! are kept. Clicking the provider's own "New chat" control is instant and
//! keeps the single-page app mounted, but it depends on that page's DOM.
//! Navigating to a new-chat URL always works but pays a full reload. The
//! frontend chooses between them and falls back; this module owns the two
//! pieces of user-supplied configuration that make either survive a provider
//! redesign — an override URL, and a description of the control to click.
//!
//! Both arrive from the webview and are therefore untrusted. The rules here
//! are about *shape*: bounded sizes, no control characters, attribute names
//! from known-safe families, and a URL that cannot leave the origin placement
//! already requires. Which attributes are worth matching on is a judgement
//! about provider markup rather than about safety, so that filtering lives in
//! the frontend parser where the paste is turned into signals.

use serde::{Deserialize, Serialize};
use tauri::Url;

use super::{
    config::Provider,
    error::{ProviderCommandError, ProviderErrorCode},
};

/// A pasted element carries a handful of short identifiers. These bounds are
/// far above anything real markup needs and far below anything that could
/// bloat the injected script.
const MAX_SIGNAL_BYTES: usize = 256;
const MAX_ATTRIBUTE_NAME_BYTES: usize = 64;
const MAX_ATTRIBUTES: usize = 8;
const MAX_URL_BYTES: usize = 2_048;

/// Attributes outside the `data-`/`aria-` families that identify a control
/// without describing how it currently looks or what state it is in.
const ALLOWED_PLAIN_ATTRIBUTES: [&str; 5] = ["href", "id", "name", "role", "type"];

fn invalid_request(message: impl Into<String>) -> ProviderCommandError {
    ProviderCommandError::new(ProviderErrorCode::InvalidRequest, message)
}

/// One `name="value"` signal extracted from the element the user pasted.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct NewChatAttribute {
    pub(crate) name: String,
    pub(crate) value: String,
}

/// A description of the provider's "New chat" control, good enough to find it
/// again after a redesign. Every field is optional because no single one
/// survives forever; the in-page scorer needs any combination that clears its
/// confidence threshold, so losing one signal degrades accuracy instead of
/// breaking the match outright.
#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct NewChatMatcher {
    #[serde(default)]
    pub(crate) test_id: Option<String>,
    #[serde(default)]
    pub(crate) label: Option<String>,
    #[serde(default)]
    pub(crate) href: Option<String>,
    #[serde(default)]
    pub(crate) attributes: Vec<NewChatAttribute>,
}

impl NewChatMatcher {
    /// Returns the matcher in canonical form, or explains why it is unusable.
    ///
    /// A matcher that survives this is still only a *hint*: the page-side
    /// scorer decides whether any element resembles it closely enough to
    /// click, so an honest-but-wrong description costs a fallback, not a
    /// misdirected click.
    pub(crate) fn sanitized(&self) -> Result<Self, ProviderCommandError> {
        let test_id = sanitize_signal(self.test_id.as_deref(), "test id")?;
        let label = sanitize_signal(self.label.as_deref(), "label")?;
        let href = sanitize_signal(self.href.as_deref(), "link target")?;

        if self.attributes.len() > MAX_ATTRIBUTES {
            return Err(invalid_request(
                "The New chat button description carries too many attributes.",
            ));
        }

        let mut attributes: Vec<NewChatAttribute> = Vec::with_capacity(self.attributes.len());
        for attribute in &self.attributes {
            let name = attribute.name.trim().to_ascii_lowercase();
            if !is_allowed_attribute_name(&name) {
                return Err(invalid_request(format!(
                    "The New chat button description contains an unsupported attribute: {name}."
                )));
            }
            let value = sanitize_signal(Some(attribute.value.as_str()), &name)?.unwrap_or_default();
            if attributes
                .iter()
                .any(|existing: &NewChatAttribute| existing.name == name)
            {
                return Err(invalid_request(format!(
                    "The New chat button description repeats the attribute {name}."
                )));
            }
            attributes.push(NewChatAttribute { name, value });
        }

        let sanitized = Self {
            test_id,
            label,
            href,
            attributes,
        };
        if sanitized.is_empty() {
            return Err(invalid_request(
                "The New chat button description has nothing to match on.",
            ));
        }
        Ok(sanitized)
    }

    fn is_empty(&self) -> bool {
        self.test_id.is_none()
            && self.label.is_none()
            && self.href.is_none()
            && self.attributes.is_empty()
    }
}

fn sanitize_signal(
    value: Option<&str>,
    field: &str,
) -> Result<Option<String>, ProviderCommandError> {
    let Some(value) = value else {
        return Ok(None);
    };
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return Ok(None);
    }
    if trimmed.len() > MAX_SIGNAL_BYTES {
        return Err(invalid_request(format!(
            "The New chat button {field} is too long."
        )));
    }
    if trimmed.chars().any(char::is_control) {
        return Err(invalid_request(format!(
            "The New chat button {field} contains invalid characters."
        )));
    }
    Ok(Some(trimmed.to_string()))
}

/// `data-`/`aria-` attributes are the families providers use for automation
/// and accessibility hooks; the short plain list covers identity attributes.
/// `class` and `style` are deliberately absent — they describe appearance,
/// change on nearly every deploy, and matching them makes an override rot
/// faster than having no override at all.
fn is_allowed_attribute_name(name: &str) -> bool {
    if name.is_empty() || name.len() > MAX_ATTRIBUTE_NAME_BYTES {
        return false;
    }
    if !name.chars().all(|character| {
        character.is_ascii_lowercase() || matches!(character, '-' | '_' | '0'..='9')
    }) {
        return false;
    }
    if name.starts_with("data-") || name.starts_with("aria-") {
        return name.len() > "data-".len();
    }
    ALLOWED_PLAIN_ATTRIBUTES.contains(&name)
}

/// Resolves where a URL-based reset should navigate.
///
/// An override may retarget the path — providers move their new-chat route,
/// and Google's `/u/<n>/` account prefix differs per person — but never the
/// origin. Placement refuses to fill a page that is not on
/// `expected_fill_host`, so an off-origin reset would strand the pane
/// somewhere the prompt can never land; rejecting it here turns that into an
/// error the user sees while editing the setting.
pub(crate) fn resolve_new_chat_url(
    provider: Provider,
    override_url: Option<&str>,
) -> Result<Url, ProviderCommandError> {
    let config = provider.config();
    let Some(candidate) = override_url
        .map(str::trim)
        .filter(|value| !value.is_empty())
    else {
        return Url::parse(config.new_chat_url).map_err(|error| {
            ProviderCommandError::new(
                ProviderErrorCode::WebviewOperationFailed,
                format!("Invalid built-in new chat URL: {error}"),
            )
        });
    };

    if candidate.len() > MAX_URL_BYTES {
        return Err(invalid_request("The New chat address is too long."));
    }
    let url = Url::parse(candidate)
        .map_err(|_| invalid_request("The New chat address is not a valid web address."))?;
    if url.scheme() != "https" {
        return Err(invalid_request(
            "The New chat address must start with https.",
        ));
    }
    if !url.username().is_empty() || url.password().is_some() {
        return Err(invalid_request(
            "The New chat address must not contain sign-in credentials.",
        ));
    }
    if url.port_or_known_default() != Some(443) {
        return Err(invalid_request(
            "The New chat address must use the standard https port.",
        ));
    }
    if url.host_str() != Some(config.expected_fill_host) {
        return Err(invalid_request(format!(
            "The New chat address must stay on {}.",
            config.expected_fill_host
        )));
    }

    Ok(url)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn matcher_with_attribute(name: &str, value: &str) -> NewChatMatcher {
        NewChatMatcher {
            attributes: vec![NewChatAttribute {
                name: name.into(),
                value: value.into(),
            }],
            ..NewChatMatcher::default()
        }
    }

    #[test]
    fn sanitizing_trims_signals_and_lowercases_attribute_names() {
        let matcher = NewChatMatcher {
            test_id: Some("  create-new-chat-button  ".into()),
            label: Some(" New chat ".into()),
            href: Some("/".into()),
            attributes: vec![NewChatAttribute {
                name: "Data-Sidebar-Item".into(),
                value: " true ".into(),
            }],
        }
        .sanitized()
        .expect("a well-formed matcher should sanitize");

        assert_eq!(matcher.test_id.as_deref(), Some("create-new-chat-button"));
        assert_eq!(matcher.label.as_deref(), Some("New chat"));
        assert_eq!(matcher.href.as_deref(), Some("/"));
        assert_eq!(matcher.attributes[0].name, "data-sidebar-item");
        assert_eq!(matcher.attributes[0].value, "true");
    }

    #[test]
    fn sanitizing_rejects_appearance_attributes_and_oversized_signals() {
        assert!(matcher_with_attribute("class", "menu-item")
            .sanitized()
            .is_err());
        assert!(matcher_with_attribute("style", "color:red")
            .sanitized()
            .is_err());
        assert!(matcher_with_attribute("onclick", "alert(1)")
            .sanitized()
            .is_err());
        assert!(matcher_with_attribute("data-", "x").sanitized().is_err());

        let oversized = NewChatMatcher {
            test_id: Some("x".repeat(MAX_SIGNAL_BYTES + 1)),
            ..NewChatMatcher::default()
        };
        assert!(oversized.sanitized().is_err());

        let control_characters = NewChatMatcher {
            label: Some("New\nchat".into()),
            ..NewChatMatcher::default()
        };
        assert!(control_characters.sanitized().is_err());
    }

    #[test]
    fn sanitizing_rejects_matchers_that_cannot_identify_anything() {
        assert!(NewChatMatcher::default().sanitized().is_err());

        let blank = NewChatMatcher {
            test_id: Some("   ".into()),
            label: Some("".into()),
            ..NewChatMatcher::default()
        };
        assert!(blank.sanitized().is_err());
    }

    #[test]
    fn sanitizing_rejects_duplicate_and_excessive_attributes() {
        let duplicated = NewChatMatcher {
            attributes: vec![
                NewChatAttribute {
                    name: "data-x".into(),
                    value: "1".into(),
                },
                NewChatAttribute {
                    name: "DATA-X".into(),
                    value: "2".into(),
                },
            ],
            ..NewChatMatcher::default()
        };
        assert!(duplicated.sanitized().is_err());

        let excessive = NewChatMatcher {
            attributes: (0..=MAX_ATTRIBUTES)
                .map(|index| NewChatAttribute {
                    name: format!("data-{index}"),
                    value: "1".into(),
                })
                .collect(),
            ..NewChatMatcher::default()
        };
        assert!(excessive.sanitized().is_err());
    }

    #[test]
    fn an_absent_override_resolves_to_the_built_in_target() {
        for provider in Provider::ALL {
            let resolved = resolve_new_chat_url(provider, None).unwrap();
            assert_eq!(resolved.as_str(), provider.config().new_chat_url);
            assert_eq!(
                resolve_new_chat_url(provider, Some("   ")).unwrap(),
                resolved
            );
        }
    }

    #[test]
    fn an_override_may_retarget_the_path_but_never_the_origin() {
        let accepted = resolve_new_chat_url(
            Provider::Gemini,
            Some("https://gemini.google.com/u/1/app/new"),
        )
        .expect("a same-origin account path should be accepted");
        assert_eq!(accepted.path(), "/u/1/app/new");

        assert!(resolve_new_chat_url(Provider::Chatgpt, Some("https://chatgpt.com/new")).is_ok());
        assert!(
            resolve_new_chat_url(Provider::Chatgpt, Some("https://chatgpt.com/?model=gpt-5"))
                .is_ok()
        );
    }

    #[test]
    fn an_override_is_held_to_the_same_origin_policy_as_placement() {
        let rejected = [
            (Provider::Chatgpt, "http://chatgpt.com/"),
            (Provider::Chatgpt, "https://chatgpt.com:8443/"),
            (Provider::Chatgpt, "https://evil.chatgpt.com/"),
            (Provider::Chatgpt, "https://chatgpt.com.evil.com/"),
            (Provider::Chatgpt, "https://gemini.google.com/app/new"),
            (Provider::Chatgpt, "javascript:alert(1)"),
            (Provider::Chatgpt, "data:text/html,<b>x</b>"),
            (Provider::Chatgpt, "https://user:pass@chatgpt.com/"),
            (Provider::Chatgpt, "not a url"),
            (Provider::Gemini, "https://chatgpt.com/"),
            (Provider::Gemini, "https://sites.google.com/view/untrusted"),
        ];
        for (provider, candidate) in rejected {
            let error = resolve_new_chat_url(provider, Some(candidate))
                .expect_err("{candidate} must be rejected");
            assert_eq!(error.code, ProviderErrorCode::InvalidRequest);
        }

        let oversized = format!("https://chatgpt.com/{}", "x".repeat(MAX_URL_BYTES));
        assert!(resolve_new_chat_url(Provider::Chatgpt, Some(&oversized)).is_err());
    }
}
